File: lib/services/notice_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/notice_model.dart';
import 'auth_service.dart';

class NoticeService {
  // Get active notices
  Future<List<Notice>> getActiveNotices({
    int? categoryId,
    String? type,
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final queryParams = <String, String>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      if (categoryId != null) {
        queryParams['categoryId'] = categoryId.toString();
      }
      if (type != null) {
        queryParams['type'] = type;
      }

      final uri = Uri.parse('${ApiConfig.baseUrl}/api/notices/active')
          .replace(queryParameters: queryParams);

      final response = await http.get(
        uri,
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final List<dynamic> noticesJson = data['notices'];
        return noticesJson.map((json) => Notice.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load notices: ${response.statusCode}');
      }
    } catch (e) {
      print('Error fetching notices: $e');
      throw Exception('Failed to load notices: $e');
    }
  }

  // Get notice by ID
  Future<Notice> getNoticeById(int id) async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/api/notices/$id'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return Notice.fromJson(data);
      } else {
        throw Exception('Failed to load notice: ${response.statusCode}');
      }
    } catch (e) {
      print('Error fetching notice: $e');
      throw Exception('Failed to load notice: $e');
    }
  }

  // Get all categories
  Future<List<NoticeCategory>> getCategories() async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/api/notice-categories'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final decoded = json.decode(response.body);
        
        // Handle both formats: direct array or wrapped in {success, data}
        final List<dynamic> data;
        if (decoded is List) {
          data = decoded;
        } else if (decoded is Map && decoded['data'] is List) {
          data = decoded['data'];
        } else {
          throw Exception('Unexpected response format');
        }
        
        return data.map((json) => NoticeCategory.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load categories: ${response.statusCode}');
      }
    } catch (e) {
      print('Error fetching categories: $e');
      throw Exception('Failed to load categories: $e');
    }
  }

  // Increment view count
  Future<void> incrementViewCount(int id) async {
    try {
      await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/notices/$id/view'),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      print('Error incrementing view count: $e');
    }
  }

  // Mark as read
  Future<void> markAsRead(int id) async {
    try {
      await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/notices/$id/read'),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      print('Error marking as read: $e');
    }
  }

  // Toggle interaction (Like, Love, RSVP)
  Future<void> toggleInteraction(int id, String type) async {
    try {
      final token = await AuthService.getAccessToken();
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/notices/$id/interact'),
        headers: {
          'Content-Type': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: json.encode({'type': type}),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to toggle interaction: ${response.statusCode}');
      }
    } catch (e) {
      print('Error toggling interaction: $e');
      throw e;
    }
  }
}

File: clean-care-admin/src/services/noticeService.ts
import axios from 'axios';
import { API_CONFIG } from '../config/apiConfig';
import {
    Notice,
    NoticeCategory,
    CreateNoticeDTO,
    UpdateNoticeDTO,
    NoticeFilters,
    NoticeListResponse,
    NoticeAnalytics,
} from '../types/notice.types';

const apiClient = axios.create({
    baseURL: API_CONFIG.BASE_URL,
    timeout: API_CONFIG.TIMEOUT,
    withCredentials: true,
    headers: {
        'Content-Type': 'application/json',
    },
});

apiClient.interceptors.request.use((config) => {
    const token = localStorage.getItem('accessToken');
    if (token) {
        config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
});

class NoticeService {
    // ==================== Notice APIs ====================

    // Get all notices (admin)
    async getAllNotices(
        filters: NoticeFilters = {},
        page = 1,
        limit = 20
    ): Promise<NoticeListResponse> {
        const params = new URLSearchParams({
            page: page.toString(),
            limit: limit.toString(),
        });

        if (filters.categoryId) params.append('categoryId', filters.categoryId.toString());
        if (filters.type) params.append('type', filters.type);
        if (filters.priority) params.append('priority', filters.priority);
        if (filters.isActive !== undefined) params.append('isActive', filters.isActive.toString());
        if (filters.search) params.append('search', filters.search);

        const response = await apiClient.get(`/api/notices?${params.toString()}`);
        return response.data;
    }

    // Get active notices (public)
    async getActiveNotices(
        filters: Partial<NoticeFilters> = {},
        page = 1,
        limit = 20
    ): Promise<NoticeListResponse> {
        const params = new URLSearchParams({
            page: page.toString(),
            limit: limit.toString(),
        });

        if (filters.categoryId) params.append('categoryId', filters.categoryId.toString());
        if (filters.type) params.append('type', filters.type);

        const response = await apiClient.get(`/api/notices/active?${params.toString()}`);
        return response.data;
    }

    // Get notice by ID
    async getNoticeById(id: number): Promise<Notice> {
        const response = await apiClient.get(`/api/notices/${id}`);
        return response.data;
    }

    // Create notice
    async createNotice(data: CreateNoticeDTO): Promise<Notice> {
        const response = await apiClient.post('/api/notices', data);
        return response.data;
    }

    // Update notice
    async updateNotice(id: number, data: UpdateNoticeDTO): Promise<Notice> {
        const response = await apiClient.put(`/api/notices/${id}`, data);
        return response.data;
    }

    // Toggle notice status
    async toggleNoticeStatus(id: number): Promise<Notice> {
        const response = await apiClient.patch(`/api/notices/${id}/toggle`);
        return response.data;
    }

    // Delete notice
    async deleteNotice(id: number): Promise<void> {
        await apiClient.delete(`/api/notices/${id}`);
    }

    // Get analytics
    async getAnalytics(): Promise<NoticeAnalytics> {
        const response = await apiClient.get('/api/notices/analytics/stats');
        return response.data;
    }

    // Increment view count
    async incrementViewCount(id: number): Promise<void> {
        await apiClient.post(`/api/notices/${id}/view`);
    }

    // Mark as read
    async markAsRead(id: number): Promise<void> {
        await apiClient.post(`/api/notices/${id}/read`);
    }

    // ==================== Category APIs ====================

    // Get all categories
    async getAllCategories(): Promise<NoticeCategory[]> {
        const response = await apiClient.get('/api/notice-categories');
        return response.data.data;
    }

    // Get category tree
    async getCategoryTree(): Promise<NoticeCategory[]> {
        const response = await apiClient.get('/api/notice-categories/tree');
        return response.data.data;
    }

    // Get category by ID
    async getCategoryById(id: number): Promise<NoticeCategory> {
        const response = await apiClient.get(`/api/notice-categories/${id}`);
        return response.data.data;
    }

    // Create category
    async createCategory(data: Partial<NoticeCategory>): Promise<NoticeCategory> {
        const response = await apiClient.post('/api/notice-categories', data);
        return response.data.data;
    }

    // Update category
    async updateCategory(id: number, data: Partial<NoticeCategory>): Promise<NoticeCategory> {
        const response = await apiClient.put(`/api/notice-categories/${id}`, data);
        return response.data.data;
    }

    // Delete category
    async deleteCategory(id: number): Promise<void> {
        await apiClient.delete(`/api/notice-categories/${id}`);
    }

    // Upload image to Cloudinary
    async uploadImage(file: File): Promise<string> {
        try {
            const formData = new FormData();
            formData.append('images', file);

            const response = await apiClient.post<{
                success: boolean;
                data: {
                    fileUrls: string[];
                };
            }>('/api/uploads', formData, {
                headers: {
                    'Content-Type': 'multipart/form-data',
                },
            });

            if (response.data.success && response.data.data.fileUrls.length > 0) {
                return response.data.data.fileUrls[0];
            }

            throw new Error('Failed to upload image');
        } catch (error) {
            console.error('Error uploading image:', error);
            throw error;
        }
    }
}

export default new NoticeService();
